### ISSUE TITLE
ThumbRestTouch output of Touch Controller node does not work properly

### ISSUE THREAD

**Issue by antrobot1234** [Yellow-Dog-Man/Resonite-Issues#449]
### Describe the bug?

The thumbRestTouch bool output of the touch controller node does not enable when you touch the thumb rest area.

### To Reproduce

open game with quest 2 (via oculus link)
spawn in touch controller node and link it to your user
touch the thumb rest
the boolean does not toggle

### Expected behavior

the boolean is supposed to toggle

### Screenshots

![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/21179158/3a0d1562-ca35-42da-aacd-b47d4855da1f)

### Resonite Version Number

10.20.831

### What Platforms does this occur on?

Windows

### What headset if any do you use?

Quest 2 Via link

### Log Files

[THETOWER-XVI- - 2023.10.20.831 - 2023-10-26 16_20_29.log](https://github.com/Yellow-Dog-Man/Resonite-Issues/files/13183021/THETOWER-XVI-.-.2023.10.20.831.-.2023-10-26.16_20_29.log)

**Comment by shadowpanther** [comment]
Works properly for me - Quest 3 via Airlink and Oculus runtime.
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/675104/64c44f4d-f871-4735-802f-a66bac9bd919)

**Comment by Nammi-namm** [comment]
I remember this issue. It would work with another user on a Rift CV1. But me on Quest 2 via airlink it wouldn't ever show true. But my avatars thumb did move when my finger went on the spot. That was a long time ago though.

**Comment by amplified1** [comment]
From my experience this issue is only present when playing on a Meta/Oculus headset through steamVR, and doesn't show up when playing directly through Oculus. This explains why it only shows up for some people, as some people use Oculus directly, and some use steamVR.

**Comment by shadowpanther** [comment]
Might it be that SteamVR just doesn't pass that input properly or is emulating the other type of controllers?

**Comment by WayaShinzui** [comment]
The thumb moves but the Quest 2 thumb rest has never worked with the node even back in LogiX. I remember having the same frustration. The avatar's thumb moves to a different position from any of the other touch buttons but does not pass a true on the node.

**Comment by amplified1** [comment]
> I saw in a SteamVR Beta update released today:
> 
> > Oculus:
> > 
> > * Hooked up thumbrest input for Touch controllers.
> 
> This sounds like it may have resolved this issue. If someone could test, and verify, this issue can be closed out. @antrobot1234 @shadowpanther @Nammi-namm @amplified1 @WayaShinzui.

I had to manually hook it up in steam bindings but it works. idk how modifying the default steamvr binds works from the developer side, but if that is fixed, this issue will be resolved

**Comment by amplified1** [comment]
Can this issue be renamed to clarify the current issue, being that the game's default steamvr bindings don't have the thumb-rest connected. @shiftyscales

**Comment by amplified1** [comment]
> Wait- I just noticed- what is the functional difference between "thumbrest" touch and "joystick" touch, @amplified1? What part of the Touch controllers are considered to be the "thumb rest"? I looked it up but I don't think my model of Touch Controllers even have that input?
> 
> I have the Quest 1 Touch Controllers.
> 
> Otherwise, yeah- I see the relevant bindings inside of SteamVR and they should probably be hooked up regardless- I'm curious about the discrepancy between models of the Oculus Touch controllers though- was the thumb rest something they added later?

the thumb rest was specifically omitted from the quest 1 controllers, but has been present in every touch controller before and after, on the quest 2 and original touch controllers it is a little circle of rough material on the face of the controller, and on quest 3 and pro it's an invisible capacitive sensor on the curved empty area of the face of the controller next to the buttons

**Comment by jeanahelver** [comment]
just stumbled on this issue a swell.
i think touch controllers should stay as a single node for touch controllers as ease of use reasons and extend a "ThumbRestForce" set to 0 for ones that don't have it, and relying on the type.

**Comment by Merith-TK** [comment]
Quest 2, 3, and Pro controller owner here
On the quest 2 the thumbrest is the discolored circle to the inside of the controller (in the flat space)
On the 3 and pro it is the slanted portion of the face (pro has an unused *mousepad like feature* that meta never exposed to the API)

The oddity of this particular spot is that the VRIK does respond to users using the thumbrest, but the Protoflux node does not.
Which is the main reason why I am annoyed that this is broked because I want to be able to setup gestures but I *cant*

**Comment by WayaShinzui** [comment]
> The oddity of this particular spot is that the VRIK does respond to users using the thumbrest, but the Protoflux node does not. Which is the main reason why I am annoyed that this is broked because I want to be able to setup gestures but I _cant_

Iirc the fingers respond because of the Steam skeletal stuff. I'm guessing Steam can see the input but it wasn't getting passed to the game?

Also I swear I read somewhere you can now manually add it to your Steam binding for Resonite and it should passthrough to the node now? I could absolutely be misremembering though and have not tested it yet myself. ADHD go brrr...

**Comment by Merith-TK** [comment]
yep, once i figure out how to share my map here to make it work i will

**Comment by Merith-TK** [comment]
@shiftyscales apologies for the direct mention
The fix is simply actually *mapping* the thumbrest_touch in the SteamVR default layout

**Comment by shiftyscales** [comment]
Sounds like this issue has everything it needs, @Frooxius @ProbablePrime - sounds like a fairly simple tweak to the default SteamVR bindings for Oculus Touch controllers.

Unfortunately I don't have later model Touch controllers that feature that input so I am unable to test it myself- but it sounds like that functionality just works after making the above recommended change.

**Comment by Merith-TK** [comment]
It is mapped, I hope to see this pushed to the default steamvr bindings soon-ish

**Comment by Merith-TK** [comment]
Are there any updates to this? It has been over a year and the fix has been stated and shown.

A community binding should not be necessary for something that can be updated without actually changing any game code (iirc doesn't steamvr store mapping data as a plain text json file?)

**Comment by Redd56** [comment]
Since we have the new community help wanted flag would this one fonstitute it since its just a json file of the bindings (to my knowledge)meaning anyone could help make new bindings?

**Comment by ko-tengu** [comment]
The bindings already exist, they just need to be updated to be the default ones, it's entirely just in the team's hands to push them.

**Comment by Merith-TK** [comment]
Has been since I opened this issue,

Sent from Proton Mail Android

-------- Original Message --------

**Comment by ko-tengu** [comment]
For the sake of making this easier, here is the updated binding Merith-TK made as the json file instead of a suggestion on where to look for it.

[steam.app.2519830_oculus_touch.json](https://github.com/user-attachments/files/19996620/steam.app.2519830_oculus_touch.json)

**Comment by Redd56** [comment]
So for anyone scrolling im gonna summerize all known info. 

The issue was originaly on steamvr. They did not have the thumbrest set up on their end. This was recitified back in march of 2024. 

Now the issue is that the default bindings for resonite do not have the thumbrest button bound for the quest 2 3 or pro controllers.

Current workarround is to use the community steamvr binding "quest 2/3/pro thumbrest support"

This is isolated to running under steam vr.

Steamvr skeletal animations are uneffected so putting your thumb on the thumbrest makes your hands move correctly.

**Comment by Merith-TK** [comment]
Other platforms such as WiVrn also don't have the thumb rest mapped by default and I don't know of whays to actually enable that. It's unmapped because it's a feature that isn't well known because nothing actually uses it.

Sent from Proton Mail Android

-------- Original Message --------

**Comment by RubberDuckShobe** [comment]
> Other platforms such as WiVrn also don't have the thumb rest mapped by default and I don't know of whays to actually enable that. It's unmapped because it's a feature that isn't well known because nothing actually uses it.

the thumbrest works with WiVRn, but you need to use a controller mapping JSON file with xrizer (or opencomposite) to get it to work

so in other words, it's unrelated to WiVRn, it's the same problem as with SteamVR - xrizer/OC uses the default OpenVR controller bindings provided by the game, which don't map the thumbrest, so it doesn't work

**Comment by Redd56** [comment]
Yeah thats what i meant i just didnt explain it well enough it seems

**Comment by Redd56** [comment]
one thing i recently learned is that the cv1, quest2, quest3, and quest pro controllers all have a thumbrest but the quest 1/rift S controllers done. i dont know why oculus decided to berid of it for the first quest then bring it back but technicly i guess this makes it to where the rifts/quest1 have to have a special 2ndary controller node.  or something else

**Comment by Redd56** [comment]
okay gonna do a quick rundown of the entire issue as is

all that needs to be done is to hook up the bindings on steamvr's end, Quest link itself works fine and the game already handles it perfectly

on linux you can modify ``Resonite/Renderer/Renderite.Renderer_Data/StreamingAssets/SteamVR/bindings_oculus_touch.json`` directly if your running on wyvrn since there is no steam overlay

for those of you wanting to manualy fix it: 
what needs to be added is on line 337 

```,
            {
               "inputs" : {
                  "touch" : {
                     "output" : "/actions/oculustouch/in/thumbrest_touch"
                  }
               },
               "mode" : "button",
               "path" : "/user/hand/left/input/thumbrest"
            },
            {
               "inputs" : {
                  "touch" : {
                     "output" : "/actions/oculustouch/in/thumbrest_touch"
                  }
               },
               "mode" : "button",
               "path" : "/user/hand/right/input/thumbrest"
            }
```
this will make the file go from looking like this: 

<img width="637" height="331" alt="Image" src="https://github.com/user-attachments/assets/1ec1635a-225f-4670-b451-3b3a2b818958" />

and will end up looking like this: 

<img width="637" height="622" alt="Image" src="https://github.com/user-attachments/assets/b48e0728-441d-47d6-a59c-26f9cffb8cad" />

once this is done the game will fully see the entirety of the thumbrest for all quest controllers that have one (quest 1 doesnt for some reason)

**Comment by Redd56** [comment]
With unity open sourced now ill try to make a pr to fix it

**Comment by Frooxius** [comment]
Merged fix in 2026.4.16.1327! Thanks!
